subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -498,52 +487,68 @@ public ServiceProvider getService(LookupContext context, TypeSpec serviceType) {
                 provider.getAll(context, serviceType, result);
             }
         }
 
-        public boolean hasService(Class<?> serviceType) {
-            return serviceTypes != null && serviceTypes.contains(serviceType);
-        }
-
         @Override
         public void stop() {
-            if (providers == null) {
-                return;
-            }
-            CompositeStoppable.stoppable(providers).stop();
+            stoppable.stop();
         }
 
         public void add(Provider provider) {
             assertMutable();
-            if (providers == null) {
-                providers = new ArrayList<Provider>();
-                serviceTypes = new HashSet<Class<?>>();
-                serviceTypes.add(Object.class);
-            }
-            this.providers.add(provider);
-            if (provider instanceof SingletonService) {
-                addServiceType(((SingletonService) provider).serviceClass);
-            } else {
+            if (!(provider instanceof SingletonService)) {
                 throw new UnsupportedOperationException("Unsupported service provider type: " + provider);
             }
+            stoppable.add(provider);
+            addProviderForClassHierarchy(((SingletonService) provider).serviceClass, provider);
         }
 
-        private void addServiceType(Class<?> serviceType) {
-            if (serviceType!=null && serviceTypes.add(serviceType)) {
-                addServiceType(serviceType.getSuperclass());
-                Class<?>[] interfaces = serviceType.getInterfaces();
-                if (interfaces!=null) {
-                    for (Class<?> intf : interfaces) {
-                        addServiceType(intf);
-                    }
+        private void addProviderForClassHierarchy(Class<?> serviceType, Provider provider) {
+            putServiceType(Object.class, provider);
+
+            Queue<Class<?>> interfaces = new ArrayDeque<Class<?>>(3);
+
+            Class<?> parent = serviceType;
+            while (parent != null && parent != Object.class) {
+                putServiceType(parent, provider);
+                Collections.addAll(interfaces, parent.getInterfaces());
+                parent = parent.getSuperclass();
+            }
+
+            if (interfaces.isEmpty()) {
+                return;
+            }
+
+            Set<Class<?>> seen = new HashSet<Class<?>>(interfaces.size());
+            while (!interfaces.isEmpty()) {
+                Class<?> next = interfaces.poll();
+                if (seen.add(next)) {
+                    putServiceType(next, provider);
+                    Collections.addAll(interfaces, next.getInterfaces());
                 }
             }
         }
+
+        private void putServiceType(Class<?> type, Provider provider) {
+            Collection<Provider> providers = providersByType.get(type);
+            if (providers == null) {
+                providers = new ArrayList<Provider>(1);
```

Comment:
:o: I wonder if it's not a bit too optimistic. There are services with lots of providers.

Reply:
JFR says this almost never grows, which makes sense since the vast majority of our interfaces has only one implementation.